### PR TITLE
Add three-state boolean problem

### PR DESCRIPTION
**Pull request by tejasbubane**
Issue: https://github.com/rubocop-hq/rubocop-rails/issues/337

Ref: https://thoughtbot.com/blog/avoid-the-threestate-boolean-problem

### CODE REVIEW

File: README.adoc [hunk 1]
```diff
@@ -1014,6 +1014,20 @@ end
 While enforcing table defaults only in Rails is suggested by many Rails developers, it's an extremely brittle approach that leaves your data vulnerable to many application bugs.
 And you'll have to consider the fact that most non-trivial apps share a database with other applications, so imposing data integrity from the Rails app is impossible.
 
+=== 3-state Boolean [[three-state-boolean]]
+
+When boolean columns are not given default value, they end up having three possible values: `true`, `false` and `NULL`.
```
Comment (bbatsov) [inline README.adoc:1019]: I guess that depends on the database, so you might want to expand on this.

File: README.adoc [hunk 2]
```diff
@@ -1014,6 +1014,20 @@ end
...
+Boolean operators work in unexpected ways with `NULL`. Avoid such situations by always adding default values to booleans columns.
```
Comment (bbatsov) [inline README.adoc:1020]: Without examples it's hard to tell what you mean by this.

File: README.adoc [hunk 3]
```diff
@@ -1014,6 +1014,20 @@ end
...
+
+[source,ruby]
+----
+# bad - boolean without default
+add_column :users, :active, :boolean
+
+# good - boolean with default
+add_column :users, :active, :boolean, default: true
```
Comment (pirj) [inline README.adoc:1028]: It might default to `true`, but it's still a three-state boolean, i.e. `User.create(active: nil).reload.active # => nil`.

Comment (tejasbubane) [reply README.adoc]: Right. I missed the `null: false` constraint. Fixed now.

File: README.adoc (hunk 2 above)
Comment (tejasbubane) [reply README.adoc]: Added example

File: README.adoc (hunk 1 above)
Comment (tejasbubane) [reply README.adoc]: Mentioned `SQL databases`

File: README.adoc [hunk 4]
```diff
@@ -1026,6 +1026,24 @@ end
 While enforcing table defaults only in Rails is suggested by many Rails developers, it's an extremely brittle approach that leaves your data vulnerable to many application bugs.
 And you'll have to consider the fact that most non-trivial apps share a database with other applications, so imposing data integrity from the Rails app is impossible.
 
+=== 3-state Boolean [[three-state-boolean]]
+
+With SQL databases, if a boolean column is not given a default value, it ends up having three possible values: `true`, `false` and `NULL`.
+Boolean operators [work in unexpected ways](https://en.wikipedia.org/wiki/Three-valued_logic) with `NULL`.
+
+eg. In SQL queries, `true AND NULL` is `NULL` (not false), `true AND NULL OR false` is `NULL` (not false). This can make SQL queries return unexpected results.
```
Comment (pirj) [inline README.adoc:1034]: `eg. In` -> `For example, in` or `E.g. in`.

File: README.adoc [hunk 5]
```diff
@@ -1026,6 +1026,24 @@ end
...
+
+To avoid such situations, boolean columns should always have a default value and not null constraint.
```
Comment (pirj) [inline README.adoc:1036]: `not null` -> `NOT NULL`.

File: README.adoc [hunk 6]
```diff
@@ -1026,6 +1026,24 @@ end
...
+
+[source,ruby]
+----
+# bad - boolean without default
+add_column :users, :active, :boolean
+
+# good - boolean with default
```
Comment (pirj) [inline README.adoc:1043]: ...and with restricted `NULL`

**Review by pirj — APPROVED** [review]
Thank you!

Added a couple of cosmetic notes.

**Comment by tejasbubane** [comment]
Thanks @pirj I have made the suggested changes.

File: README.adoc [hunk 7]
```diff
@@ -1026,6 +1026,24 @@ end
...
+E.g. in SQL queries, `true AND NULL` is `NULL` (not false), `true AND NULL OR false` is `NULL` (not false). This can make SQL queries return unexpected results.
+
+To avoid such situations, boolean columns should always have a default value and `NOT NULL` constraint.
+
+[source,ruby]
+----
+# bad - boolean without default
+add_column :users, :active, :boolean
+
+# good - boolean with default and restricted with null
```
Comment (pirj) [inline README.adoc:1043]: Sorry for nitpicking, but I'm not too confident if "restricted with null" explains what `null: false` does.
I think `with restricted NULL`. A quick search over https://guides.rubyonrails.org/active_record_migrations.html didn't yield any better wording, unfortunately.

Comment (tejasbubane) [reply README.adoc:1043]: changed

File: README.adoc [hunk 8]
```diff
@@ -1026,6 +1026,24 @@ end
 While enforcing table defaults only in Rails is suggested by many Rails developers, it's an extremely brittle approach that leaves your data vulnerable to many application bugs.
 And you'll have to consider the fact that most non-trivial apps share a database with other applications, so imposing data integrity from the Rails app is impossible.
 
+=== 3-state Boolean [[three-state-boolean]]
+
+With SQL databases, if a boolean column is not given a default value, it ends up having three possible values: `true`, `false` and `NULL`.
+Boolean operators [work in unexpected ways](https://en.wikipedia.org/wiki/Three-valued_logic) with `NULL`.
+
+E.g. in SQL queries, `true AND NULL` is `NULL` (not false), `true AND NULL OR false` is `NULL` (not false). This can make SQL queries return unexpected results.
```
Comment (andyw8) [inline README.adoc:1034]: Minor, but starting a sentence with "E.g." feels awkward. How about "For example"?

File: README.adoc [hunk 9]
```diff
@@ -1026,6 +1026,24 @@ end
 While enforcing table defaults only in Rails is suggested by many Rails developers, it's an extremely brittle approach that leaves your data vulnerable to many application bugs.
 And you'll have to consider the fact that most non-trivial apps share a database with other applications, so imposing data integrity from the Rails app is impossible.
 
+=== 3-state Boolean [[three-state-boolean]]
+
+With SQL databases, if a boolean column is not given a default value, it ends up having three possible values: `true`, `false` and `NULL`.
```
Comment (andyw8) [inline README.adoc:1031]: I think "will have" rather than "ends up having" reads better.

File: README.adoc [hunk 10]
```diff
@@ -1026,6 +1026,24 @@ end
...
+Boolean operators [work in unexpected ways](https://en.wikipedia.org/wiki/Three-valued_logic) with `NULL`.
+
+E.g. in SQL queries, `true AND NULL` is `NULL` (not false), `true AND NULL OR false` is `NULL` (not false). This can make SQL queries return unexpected results.
+
+To avoid such situations, boolean columns should always have a default value and `NOT NULL` constraint.
```
Comment (andyw8) [inline README.adoc:1036]: "and **a** `NOT NULL`"

**Review by andyw8 — APPROVED** [review]

File: README.adoc [hunk 11]
```diff
@@ -1026,6 +1026,24 @@ end
...
+
+[source,ruby]
+----
+# bad - boolean without a default value
+add_column :users, :active, :boolean
+
+# good - boolean with a default value and with restricted NULL
+add_column :users, :active, :boolean, default: true, null: false
```
Comment (koic) [inline README.adoc:1044]: I'd prefer an example code of the rule that defaults to `false` here :-) Of course, In practice, `false` or `true` is chosen depending on the context.

File: README.adoc [hunk 12]
```diff
@@ -1026,6 +1026,24 @@ end
...
+E.g. in SQL queries, `true AND NULL` is `NULL` (not false), `true AND NULL OR false` is `NULL` (not false). This can make SQL queries return unexpected results.
+
+To avoid such situations, boolean columns should always have a default value and `NOT NULL` constraint.
+
+[source,ruby]
+----
+# bad - boolean without a default value
+add_column :users, :active, :boolean
+
+# good - boolean with a default value and with restricted NULL
```
Comment (koic) [inline README.adoc:1043]:
```suggestion
# good - boolean with a default value (`false` or `true`) and with restricted NULL
```

File: README.adoc (hunk 11 above)
Comment (koic) [reply README.adoc:1044]: Or I think one approach is to exemplify an example with `default: false` and an example with `default: true`.

**Comment by tejasbubane** [comment]
@andyw8 @koic Thanks. I have made the suggested changes.

**Comment by koic** [comment]
Thanks!

**Pull request merged**